File: algorithms/0206-reverse-linked-list/Solution.kt
// Reverse Linked List — iterative three-pointer reversal (prev, cur, next).
// Time O(n), Space O(1).

class ListNode(var `val`: Int, var next: ListNode? = null)

class Solution {
    fun reverseList(head: ListNode?): ListNode? {
        var prev: ListNode? = null
        var cur = head
        while (cur != null) {
            val next = cur.next
            cur.next = prev
            prev = cur
            cur = next
        }
        return prev
    }
}

File: algorithms/0206-reverse-linked-list/solution.py
# Reverse Linked List - iterative three-pointer reversal (prev, cur, next).
# Time O(n), Space O(1).

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseList(self, head: ListNode | None) -> ListNode | None:
        prev = None
        cur = head
        while cur is not None:
            nxt = cur.next
            cur.next = prev
            prev = cur
            cur = nxt
        return prev
